crates/circuit/src/dag_circuit.rs

```diff
@@ -6277,6 +6241,51 @@ impl DAGCircuit {
             Err(DAGCircuitError::new_err("Specified node is not an op node"))
         }
     }
+
+    pub fn count_ops(
+        &self,
+        py: Python,
+        recurse: bool,
+    ) -> PyResult<IndexMap<String, usize, RandomState>> {
+        if !recurse || !self.has_control_flow() {
+            Ok(self.op_names.clone())
+        } else {
+            fn inner(
+                py: Python,
+                dag: &DAGCircuit,
+                counts: &mut IndexMap<String, usize, RandomState>,
+            ) -> PyResult<()> {
+                for (key, value) in dag.op_names.iter() {
+                    counts
+                        .entry(key.clone())
+                        .and_modify(|count| *count += value)
+                        .or_insert(*value);
+                }
+                let circuit_to_dag = imports::CIRCUIT_TO_DAG.get_bound(py);
```

Comment:
Maybe it's worth adding a `count_ops` method to `CircuitData`, and calling that directly instead of converting to a `DAGCircuit` first.

I'm not saying that we should necessarily bother maintaining op counts in `CircuitData` like we do in a `DAGCircuit`, but computing on the fly should still be faster than recursively converting everything to a `DAGCircuit`.

Not necessary for this PR, but I wouldn't be opposed to doing it here either.

Reply:
It's funny you mention this, that was added last week: https://github.com/Qiskit/qiskit/commit/5fc1635b6c447e910dbf9f390eb1e6ec789e70a4

but the implementation there is governed by the api of the python space method and not how I would implement it for performance in the transpiler.